text/0000-default-field-values.md

```diff
@@ -0,0 +1,1767 @@
+
+## Field default values
+
+### Grammar
+
+Let the grammar of record fields in `struct`s and `enum` variants be defined
+like so (in the `.lyg` notation):
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type;
+```
+
+Then, `RecordField` is changed into:
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type { "=" def:Expr }?;
+```
+
+Further, given the following partial definition for the expression grammar:
+
+```rust
+Expr = attrs:OuterAttr* kind:ExprKind;
+ExprKind =
+  | ...
+  | Struct:{ path:Path "{" attrs:InnerAttr* fields:StructExprFieldsAndBase "}" }
+  ;
+
+StructExprFieldsAndBase =
+  | Fields:{ fields:StructExprField* % "," ","? }
+  | Base:{ ".." base:Expr }
+  | FieldsAndBase:{ fields:StructExprField+ % "," "," ".." base:Expr }
+  ;
+StructExprField = attrs:OuterAttr* kind:StructExprFieldKind;
+StructExprFieldKind =
+  | Shorthand:IDENT
+  | Explicit:{ field:FieldName ":" expr:Expr }
+  ;
+```
+
+the rule `StructExprFieldsAndBase` is extended with:
+
+```rust
+StructExprFieldsAndBase =| FieldsAndDefault:{ fields:StructExprField+ % "," "," ".." };
+```
+
+### Static semantics
+
+#### Defining defaults
+
+Given a `RecordField` where the default is specified, i.e.:
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type "=" def:Expr;
+```
+
+all the following rules apply when type-checking:
+
+1. The expression `def` must be a constant expression.
+
+2. The expression `def` must unify with the type `ty`.
+
+When lints check attributes such as `#[allow(lint_name)]` are placed on a
+`RecordField`, it also applies to `def` if it exists.
+
+#### Initialization expressions
+
+Given an expression `τ` of the form (e.g. `Foo { x: 1, .. }`):
+
+```rust
+path:Path "{" attrs:InnerAttr* fields:StructExprField+ % "," "," ".." "}"
+```
+
+all the following rules apply when type-checking:
+
+1. All `fields` in `τ` mentioned must exist in the `struct` or `enum` variant
+   referenced by `path`. No field in `fields` may be mentioned twice.
+
+2. For each `field: expr` pair in `fields` of `τ`,
+   `expr` must unify with the type of `field` in the `struct`
+   or `enum` variant referenced by `path`.
+
+3. For each `field: expression` pair in `fields` of `τ`,
+   `field` must be visible in the context which `τ` occurs in.
+
+4. For each field `f` in `τ` *not* mentioned in `fields`, a default value must
+   exist for `f` in the `struct` or `enum` variant referenced by `path`.
+
+---------------------
+
+**Lemma 1:** If `τ` is of a type `σ` where `σ: Sync`, and if all `field: expr`
+has an `expr` component which is a valid constant expression, then `τ` is
+also a valid constant expression.
+
+**Proof:** It trivially holds that if all `expr`s in `τ` are
+constant expressions and all default values for fields omitted
+are also constant expressions, then `τ` is a constant expression.
+
+**Lemma 2:** If an expression `x` of form `Path { .. }` is of a type `σ` where
```

Comment:
If this stays in the RFC (I agree with @compiler-errors that it probably shouldn't), these two cases can be unified by letting `τ` have zero or more fields specified rather than one or more.